lib/cn/vblock_builder.c

```diff
@@ -15,24 +18,110 @@
 #include "cn_mblocks.h"
 #include "cn_metrics.h"
 #include "cn_perfc.h"
 
-#include <mpool/mpool.h>
+#define WBUF_LEN_MAX      ((1024 * 1024) + VBLOCK_FOOTER_LEN)
+
+/**
+ * struct vblock_builder - create vblocks from a stream of values
+ * @ds:        mpool dataset
+ * @pc:        performance counters
+ * @vblk_list: list of vblocks
+ * @wbuf:      write buffer
+ * @wbuf_off:  offset of next unused byte in write buffer
+ * @wbuf_len:  length of next write to media
+ * @vblk_off:  offset of next unused byte in vblock
+ * @vsize:     vblock size for compaction stats.  for vblocks, vsize
+ *             is the number of bytes written to the vblock before committing it
+ * @destruct:  if true, vlbock builder is ready to be destroyed
+ * @mblocksz:  mblock size of specified media class
+ *
+ * @cur_kobj:  keyobj context for the current vbb_add_entry()
+ * @cur_minklen: min key length
+ * @cur_minkey:  a copy of the min key referencing this vblock
+ *
+ * WBUF_LEN_MAX is the allocated size of the write buffer.  Each mblock write
+ * will be at most WBUF_LEN_MAX bytes.  Member @wbuf_len is the actual write
+ * size, and is <= WBUF_LEN_MAX.
+ *
+ * The vblock builder creates as many vblocks as needed to store the values.
+ * The write buffer is allocated once when the builder is created, and is
+ * reused between vblocks.  The following logic explains how the vlbock
+ * builder state is managed as new values are added.
+ *
+ * When a new value is given to the vblock builder
+ * -----------------------------------------------
+ *
+ *   Let @vlen be the length of the new value
+ *
+ *   If current vblock has not been allocated, start a new vblock as follows:
+ *     - allocate vblock
+ *     - set @wbuf_len to WBUF_LEN_MAX - VBLOCK_FOOTER_LEN, reserving space for footer
+ *
+ *   If current vblock does not have room for new value:
+ *     - write residual contents of @wbuf to mblock
+ *     - start a new vblock as described above
+ *
+ *   While @vlen > 0:
+ *     - copy whatever fits into @wbuf (cannot exceed @wbuf_len)
+ *     - let @copied be number of bytes copied
+ *     - set @vlen -= @copied
+ *     - set @wbuf_off += @copied
+ *     - if @wbuf_off == @wbuf_len:
+ *       -- write @wbuf_len bytes to mblock
+ *       -- set @wbuf_off to 0
+ *       -- set @vblk_off += @wbuff_off
+ */
+struct vblock_builder {
+    struct mpool *             ds;
+    struct cn *                cn;
+    struct perfc_set *         pc;
+    struct cn_merge_stats *    mstats;
+    struct blk_list            vblk_list;
+    enum hse_mclass_policy_age agegroup;
+    uint64_t                   vsize;
+    uint64_t                   blkid;
+    uint32_t                   max_size;
+    off_t                      vblk_off;
+    void *                     wbuf;
+    off_t                      wbuf_off;
+    unsigned int               wbuf_len;
+    uint64_t                   vgroup;
+    bool                       destruct;
+    const struct key_obj      *cur_kobj;
```

Comment:
Does it work to not store cur_kobj in this struct and pass first and last key as params into vblock_start and vblock_finish?  I think that woud be cleaner because we wouldn't be saving ptrs to callers memory and wouldn't have to update cur_kobj on every call to vbb_add_entry.

Reply:
You are absolutely correct! Passing it as a param is much cleaner. I started in that direction first but later added it to the builder for no good reason. Thanks!